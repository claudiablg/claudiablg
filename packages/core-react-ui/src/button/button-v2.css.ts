import { Theme } from '../design-system';
import { globalThemeReversedSelector } from '../global/global-theme-classnames';

const iconSize = `1.5em`;
const iconPadding = `3em`;

/**
 * Base
 */

import { style } from '@vanilla-extract/css';
import { calc } from '@vanilla-extract/css-utils';
import { themeVars } from '../theme/default-theme.css';

export const base = style({
  display: 'block',
  position: 'relative',
  outline: 'none',
  lineHeight: 0,
  cursor: 'pointer',
  appearance: 'none',
  userSelect: 'none',
  touchAction: 'none',
  WebkitAppearance: 'none',
  whiteSpace: 'nowrap', // don't allow wrapping
  borderRadius: 4,
  width: 'fit-content',
  ':focus': {
    // boxShadow: `0 0 0 1px ${cssTheme.colors.colors.primary[700]}`,
  },
});

/**
 * States
 */
export const rest = style({});
export const loading = style({});
export const disabled = style({});

/**
 *
 * Variants
 *
 */

/**
 *
 * Sizes
 *
 */

export const large = style({
  padding: themeVars.buttons.medium.padding.default,
});

export const medium = style({
  padding: themeVars.buttons.medium.padding.default,
});

export const small = style({
  padding: themeVars.buttons.medium.padding.default,
});

export const xSmall = style({
  padding: themeVars.buttons.medium.padding.default,
});
