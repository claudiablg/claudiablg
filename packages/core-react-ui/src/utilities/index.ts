export * from './colors.utilities';
export * from './component.utilities';
export * from './css-variable.utilities';
export * from './effects.utilities';
export * from './filter-color.utility';
export * from './text.utilities';
export * from './theme.utilities';
export * from './utilities';
