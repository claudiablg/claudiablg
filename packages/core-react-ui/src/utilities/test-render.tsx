import * as React from 'react';
import { render } from '@testing-library/react';
import { TreatProvider } from 'react-treat/lib/types';

export function renderTest(Component: any) {
  return render(<TreatProvider theme={'light'}>{Component}</TreatProvider>);
}
