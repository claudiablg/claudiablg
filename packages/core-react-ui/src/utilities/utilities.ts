import { ColorIntents, Colors } from '@newrade/core-design-system';

/**
 * Create default color intents.
 */
export function createDefaultColorIntents(colors: Colors): ColorIntents {
  return {
    primary: colors.primary500,
    primaryReversed: colors.greyscale0,
    secondary: colors.primary500,
    secondaryReversed: colors.greyscale0,
    primaryText: colors.greyscale900,
    primaryTextReversed: colors.greyscale0,
    secondaryText: colors.greyscale700,
    secondaryTextReversed: colors.greyscale0,
    tertiaryText: colors.greyscale600,
    tertiaryTextReversed: colors.greyscale0,
    successText: colors.utilityGreen1000,
    successAction: colors.utilityGreen500,
    successBackground: colors.utilityGreen100,
    warningText: colors.utilityYellow1000,
    warningAction: colors.utilityYellow500,
    warningBackground: colors.utilityYellow100,
    dangerText: colors.utilityRed1000,
    dangerAction: colors.utilityRed500,
    dangerBackground: colors.utilityRed100,
    background0: colors.greyscale0,
    background1: colors.greyscale100,
    background2: colors.greyscale200,
  };
}

/**
 * Convert a number (in px) into a rem value.
 * @param px value in pixel
 * @param baseUnitPx font-size set on the <html/> element
 */
export function pxToRem({ baseUnitPx, value }: { baseUnitPx: number; value: number }): string {
  return `${value / baseUnitPx}rem`;
}

/**
 * Convert a number (in px) into a em relative value.
 * @param px value in pixel
 * @param fontSize font-size set on the element
 */
export function pxToEm({ fontSize, value }: { fontSize: number; value: number }): string {
  return `${value / fontSize}em`;
}

/**
 * Convert px value (e.g. '15px') into a rem value.
 * @param px value in string px value
 * @param baseUnitPx font-size set on the <html/> element
 */
export function pxStringToRem({ baseFontSize, value }: { baseFontSize: number; value: string }): string {
  if (!value?.length) {
    return '';
  }

  const pxPattern = /(\d+\.?(\d+))/g;
  const match = pxPattern.exec(value);
  if (!(match?.length && match[0]?.length)) {
    return '';
  }
  const conversion = Number(match[0]);
  if (!conversion) {
    return '';
  }

  return `${conversion / baseFontSize}rem`;
}

/**
 * Convert px value (e.g. '15px') into a number value.
 * @param px value in string px value
 */
export function pxStringToNumber({ value }: { value: string }): number | undefined {
  if (!value?.length) {
    return undefined;
  }

  const pxPattern = /(\d+\.?(\d+))/g;
  const match = pxPattern.exec(value);
  if (!(match?.length && match[0]?.length)) {
    return undefined;
  }
  const conversion = Number(match[0]);
  if (conversion === undefined || conversion === NaN || conversion === null) {
    return undefined;
  }

  return conversion;
}
