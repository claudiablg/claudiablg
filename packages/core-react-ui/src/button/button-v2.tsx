import {
  ButtonIcon,
  ButtonProps,
  ButtonSize,
  LABEL_SIZE,
  TEXT_STYLE,
} from '@newrade/core-design-system';
import React, { ButtonHTMLAttributes, useRef } from 'react';
import { usePreventPinchZoom } from '../hooks/use-prevent-pinch-zoom';
import { CommonComponentProps } from '../props/component-common.props';
import { Label } from '../text/label';
import { getDefaultTextFromProps, getMergedClassname } from '../utilities/component.utilities';
import * as styles from './button-v2.css';

type Props = CommonComponentProps &
  ButtonHTMLAttributes<any> &
  Pick<ButtonProps, 'icon' | 'role' | 'size' | 'state' | 'variant'> & {
    as?: 'button' | 'a' | 'div';
  } & {
    disabled?: boolean;
    loading?: boolean;
    Icon?: React.ReactNode;
    dataPressed?: boolean;
    collapsePadding?: 'left' | 'right';
  };

/**
 * General purpose button element, can be rendered using the AsElement={} prop.
 * @version 2
 */
export const ButtonV2 = React.forwardRef<any, Props>(
  (
    {
      id,
      style,
      className,
      children,
      variant,
      collapsePadding,
      as,
      AsElement,
      size,
      state,
      disabled,
      icon = ButtonIcon.right,
      Icon,
      ...props
    },
    forwardedRef
  ) => {
    const localRef = useRef<HTMLButtonElement>(null);
    const ref = forwardedRef ? (forwardedRef as React.RefObject<HTMLButtonElement>) : localRef;
    const type = as ? as : 'button';

    /**
     *
     * Icon
     *
     */

    const iconClassNames = getMergedClassname([
      // styles.base,
      // icon === ButtonIcon.right ? styles.right : '',
      // icon === ButtonIcon.left ? styles.left : '',
      // icon === ButtonIcon.icon ? styles.icon : '',
    ]);

    const IconSvg =
      icon && icon !== ButtonIcon.none && Icon
        ? React.cloneElement(Icon as React.ReactElement, {
            className: iconClassNames,
            preserveAspectRatio: `xMinYMin meet`,
          })
        : null;

    /**
     * Event handling
     */
    usePreventPinchZoom(ref.current);

    /**
     * Styles
     */
    const classNames = getMergedClassname([
      className,
      styles.base,
      styles[size ? size : ButtonSize.medium],
      // styles.primary,
      // styles[variant ? variant : Variant.primary],
    ]);
    const renderedChildren = children
      ? children
      : getDefaultTextFromProps('button', {
          variant,
          size,
          icon,
          disabled,
        });

    function getLabelSizeForButtonSize(size?: ButtonSize): LABEL_SIZE {
      switch (size) {
        case ButtonSize.large: {
          return LABEL_SIZE.medium;
        }
        case ButtonSize.medium: {
          return LABEL_SIZE.small;
        }
        default:
        case ButtonSize.small: {
          return LABEL_SIZE.small;
        }
        case ButtonSize.xSmall: {
          return LABEL_SIZE.xSmall;
        }
      }
    }

    const CustomElement = AsElement
      ? React.cloneElement(
          AsElement as React.ReactElement,
          {
            id,
            style,
            className: classNames,
            ref: ref,
            dataicon: `${icon}`,
            datapaddingcollapse: `${collapsePadding}`,
            ...props,
          },
          <>
            {icon === ButtonIcon.icon ? null : (
              <Label
                variantDisplay={'inline'}
                variantStyle={TEXT_STYLE.bold}
                variant={getLabelSizeForButtonSize(size)}
              >
                {renderedChildren}
              </Label>
            )}
            {IconSvg}
          </>
        )
      : null;

    if (CustomElement) {
      return CustomElement;
    }

    return (
      <button
        id={id}
        style={style}
        className={classNames}
        ref={ref}
        disabled={disabled}
        // @ts-ignore
        dataicon={`${IconSvg ? icon : ''}`}
        datapaddingcollapse={`${collapsePadding}`}
        {...props}
      >
        {icon === ButtonIcon.icon ? null : (
          <Label
            variantDisplay={'inline'}
            variantStyle={TEXT_STYLE.bold}
            variant={getLabelSizeForButtonSize(size)}
          >
            {renderedChildren}
          </Label>
        )}
        {IconSvg}
      </button>
    );
  }
);
