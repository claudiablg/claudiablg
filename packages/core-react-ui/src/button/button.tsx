import {
  ButtonIcon,
  ButtonProps,
  ButtonSize,
  ButtonState,
  ButtonVariant,
  LABEL_SIZE,
  TEXT_STYLE,
} from '@newrade/core-design-system';
import { AriaButtonProps } from '@react-types/button';
import React, { useRef } from 'react';
import { useButton } from 'react-aria';
import { useStyles } from 'react-treat';
import { usePreventPinchZoom } from '../';
import { CommonComponentProps } from '../props/component-common.props';
import { Label } from '../text/label';
import { getDefaultTextFromProps, getMergedClassname } from '../utilities/component.utilities';
import * as stylesRef from './button.treat';

type Props = CommonComponentProps &
  Pick<ButtonProps, 'icon' | 'role' | 'size' | 'state' | 'variant'> &
  AriaButtonProps & { as?: 'button' | 'a' | 'div' } & {
    loading?: boolean;
    Icon?: React.ReactNode;
    dataPressed?: boolean;
    collapsePadding?: 'left' | 'right';
  };

export const Button = React.forwardRef<any, Props>(
  (
    {
      id,
      style,
      className,
      children,
      variant,
      collapsePadding,
      as,
      AsElement,
      size,
      state,
      icon = ButtonIcon.right,
      Icon,
      ...props
    },
    forwardedRef
  ) => {
    const { styles } = useStyles(stylesRef);
    const localRef = useRef<HTMLButtonElement>(null);
    const ref = forwardedRef ? (forwardedRef as React.RefObject<HTMLButtonElement>) : localRef;
    const type = as ? as : 'button';
    const { buttonProps, isPressed } = useButton({ ...props, elementType: type }, ref as any);

    const iconClassNames = getMergedClassname([styles.iconBase, icon ? styles[icon] : '']);

    usePreventPinchZoom(ref.current);
    // usePreventLongPress(ref.current);

    const IconSvg =
      icon && icon !== ButtonIcon.none && Icon
        ? React.cloneElement(Icon as React.ReactElement, {
            className: iconClassNames,
            preserveAspectRatio: `xMinYMin meet`,
          })
        : null;

    const variantStateClassName = `${styles[ButtonState.rest]}`;
    const variantClassName = `${styles[variant ? variant : ButtonVariant.primary]}`;
    const variantSizeClassName = styles[size ? size : ButtonSize.medium];
    const activeClassName = isPressed ? styles.pressed : '';
    const allClassName = getMergedClassname([
      variantStateClassName,
      variantSizeClassName,
      variantClassName,
      activeClassName,
      className,
    ]);
    const renderedChildren = children
      ? children
      : getDefaultTextFromProps('button', {
          variant,
          size,
          icon,
          disabled: props.isDisabled,
        });

    function getLabelSizeForButtonSize(size?: ButtonSize): LABEL_SIZE {
      switch (size) {
        case ButtonSize.large: {
          return LABEL_SIZE.medium;
        }
        case ButtonSize.medium: {
          return LABEL_SIZE.small;
        }
        default:
        case ButtonSize.small: {
          return LABEL_SIZE.small;
        }
        case ButtonSize.xsmall: {
          return LABEL_SIZE.xSmall;
        }
      }
    }

    const CustomElement = AsElement
      ? React.cloneElement(
          AsElement as React.ReactElement,
          {
            id,
            style,
            className: allClassName,
            ref: ref,
            ...buttonProps,
            dataicon: `${icon}`,
            datapressed: `${isPressed}`,
            datapaddingcollapse: `${collapsePadding}`,
          },
          <>
            {icon === ButtonIcon.icon ? null : (
              <Label
                style={{ display: 'inline-block' }}
                variantStyle={TEXT_STYLE.bold}
                variant={getLabelSizeForButtonSize(size)}
              >
                {renderedChildren}
              </Label>
            )}
            {IconSvg}
          </>
        )
      : null;

    if (CustomElement) {
      return CustomElement;
    }

    return (
      <button
        id={id}
        style={style}
        className={allClassName}
        ref={ref}
        {...buttonProps}
        // @ts-ignore
        dataicon={`${IconSvg ? icon : ''}`}
        datapressed={`${isPressed}`}
        datapaddingcollapse={`${collapsePadding}`}
      >
        {icon === ButtonIcon.icon ? null : (
          <Label
            style={{ display: 'inline-block' }}
            variantStyle={TEXT_STYLE.bold}
            variant={getLabelSizeForButtonSize(size)}
          >
            {renderedChildren}
          </Label>
        )}
        {IconSvg}
      </button>
    );
  }
);
