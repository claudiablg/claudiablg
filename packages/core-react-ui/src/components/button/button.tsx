import {
  ButtonIcon,
  ButtonProps,
  ButtonSize,
  ButtonState,
  ButtonVariant,
  LABEL_SIZE,
  TEXT_STYLE,
} from '@newrade/core-design-system';
import { AriaButtonProps } from '@react-types/button';
import React, { useRef, useEffect } from 'react';
import { useButton } from 'react-aria';
import { IoAddOutline } from 'react-icons/io5';
import { useStyles } from 'react-treat';
import { CommonComponentProps } from '../../props/component-common.props';
import { getDefaultTextFromProps, getMergedClassname } from '../../utilities/component.utilities';
import { Label } from '../text/label';
import * as stylesRef from './button.treat';
import { usePreventPinchZoom, usePreventLongPress } from '../..';

type Props = CommonComponentProps &
  Pick<ButtonProps, 'icon' | 'role' | 'size' | 'state' | 'variant'> &
  AriaButtonProps & { as?: 'button' | 'a' | 'div' } & {
    loading?: boolean;
    Icon?: React.ReactNode;
    dataPressed?: boolean;
    collapsePadding?: 'left' | 'right';
  };

export const Button = React.forwardRef<any, Props>(
  (
    {
      id,
      style,
      className,
      children,
      variant,
      collapsePadding,
      as,
      AsElement,
      size,
      state,
      icon = ButtonIcon.right,
      Icon,
      ...props
    },
    ref
  ) => {
    const { styles } = useStyles(stylesRef);
    const refLocal = ref ? (ref as React.RefObject<HTMLButtonElement>) : useRef<HTMLButtonElement>(null);
    const type = as ? as : 'button';
    const { buttonProps, isPressed } = useButton({ ...props, elementType: type }, refLocal as any);

    const iconClassNames = getMergedClassname([styles.iconBase, icon ? styles[icon] : '']);

    usePreventPinchZoom(refLocal.current);
    usePreventLongPress(refLocal.current);

    const IconSvg =
      Icon && icon ? (
        React.cloneElement(Icon as React.ReactElement, {
          className: iconClassNames,
          preserveAspectRatio: `xMinYMin meet`,
        })
      ) : icon ? (
        <IoAddOutline className={iconClassNames} preserveAspectRatio="xMinYMin meet" />
      ) : null;

    const variantStateClassName = `${styles[ButtonState.rest]}`;
    const variantClassName = `${styles[variant ? variant : ButtonVariant.primary]}`;
    const collapsePaddingClassName = `${collapsePadding === 'left' ? styles.collapsePaddingLeft : ''}`;
    const variantSizeClassName = styles[size ? size : ButtonSize.medium];
    const activeClassName = isPressed ? styles.pressed : '';
    const allClassName = getMergedClassname([
      variantStateClassName,
      variantSizeClassName,
      variantClassName,
      activeClassName,
      className,
    ]);
    const renderedChildren = children
      ? children
      : getDefaultTextFromProps('button', {
          variant,
          size,
          icon,
          disabled: props.isDisabled,
        });

    function getLabelSizeForButtonSize(size?: ButtonSize): LABEL_SIZE {
      switch (size) {
        case ButtonSize.large: {
          return LABEL_SIZE.medium;
        }
        case ButtonSize.medium: {
          return LABEL_SIZE.small;
        }
        default:
        case ButtonSize.small: {
          return LABEL_SIZE.small;
        }
        case ButtonSize.xsmall: {
          return LABEL_SIZE.xSmall;
        }
      }
    }

    const CustomElement = AsElement
      ? React.cloneElement(
          AsElement as React.ReactElement,
          {
            id,
            style,
            className: allClassName,
            ref: refLocal,
            ...buttonProps,
            dataicon: `${icon}`,
            datapressed: `${isPressed}`,
            datapaddingcollapse: `${collapsePadding}`,
          },
          <>
            {icon === ButtonIcon.icon ? null : (
              <Label
                style={{ display: 'inline-block' }}
                variantStyle={TEXT_STYLE.bold}
                variant={getLabelSizeForButtonSize(size)}
              >
                {renderedChildren}
              </Label>
            )}
            {IconSvg}
          </>
        )
      : null;

    if (CustomElement) {
      return CustomElement;
    }

    return (
      <button
        id={id}
        style={style}
        className={allClassName}
        ref={refLocal}
        {...buttonProps}
        // @ts-ignore
        dataicon={`${icon}`}
        datapressed={`${isPressed}`}
        datapaddingcollapse={`${collapsePadding}`}
      >
        {icon === ButtonIcon.icon ? null : (
          <Label
            style={{ display: 'inline-block' }}
            variantStyle={TEXT_STYLE.bold}
            variant={getLabelSizeForButtonSize(size)}
          >
            {renderedChildren}
          </Label>
        )}
        {IconSvg}
      </button>
    );
  }
);
