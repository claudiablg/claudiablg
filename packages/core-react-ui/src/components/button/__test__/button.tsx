import React from 'react';
import { renderTest } from '../../../utilities/test-render';
import { Button } from '../button';

describe('props', () => {
  it('should render correctly', () => {
    const { container } = renderTest(<Button>Hello world</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });
});
