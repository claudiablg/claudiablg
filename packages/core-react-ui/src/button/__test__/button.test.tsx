import React from 'react';
import { renderTest } from '../../utilities/test-render';
import { Button } from '../button';

describe('the button component', () => {
  it('should render correctly', () => {
    const { container } = renderTest(<Button>Hello world</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });
});
