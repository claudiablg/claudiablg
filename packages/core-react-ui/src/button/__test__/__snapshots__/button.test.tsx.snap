// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the button component should render correctly 1`] = `
<button
  class="style_theme style_theme style_theme"
  dataicon="none"
>
  <div
    class="style_theme style_theme style_theme style_theme"
  >
    Hello world
  </div>
</button>
`;
